docs: first draft of documentation for call spread (#381)
Agoric/documentation: main/zoe/guide/contracts/callSpread.md

# Call Spread

<Zoe-Version/>

##### [View the code on Github](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/callSpread.js) (Last updated: 03-NOV-2020)
##### [View all contracts on Github](https://github.com/Agoric/agoric-sdk/tree/master/packages/zoe/src/contracts)

> This contract implements a fully collateralized call spread. A call spread can be used as a

Reviewer: ```suggestion
This contract implements a fully collateralized call spread. You can use a call spread as a
```

financial building block to create futures, puts, calls, and event binaries that would form the

> basis for a prediction market, insurance, and much more. A Call Spread is a combination of a call

Reviewer: Should be consistent on "Call Spread" capitalization; previously on line 8 you had "call spread".

> option bought at one strike price and a second call option sold at a higher price. That means there
> are two parties to the call spread, each holding an ERTP asset that pays out complementary amounts

Reviewer: ```suggestion
option bought at one strike price and a second call option sold at a higher price. The call spread has
two parties, each holding an ERTP asset that pays out complementary amounts
```

depending on the value of some good at a known future time.

> The Zoe invitations representing the options are produced in pairs that represent the matching

Reviewer: ```suggestion
The Zoe invitations representing the options are produced in pairs representing the matching
```

assets.  The purchaser deposits the entire amount that will be approtioned between the holders of

> the two positions. The individual options are ERTP invitations that are suitable for resale.

Reviewer: ```suggestion
the two positions. The individual options are ERTP invitations that can be resold.
```

> This option is settled financially. There is no requirement that the original purchaser have

Reviewer: Which option? You've been talking about pairs of them.

ownership of the underlying asset at the start, and the beneficiaries shouldn't expect to take
delivery at closing.

> It is common for the strike price and collateral to use the same currency, though this contract
> decouples them. This would be like a spread based on the price of APPL stock (`Underlying`),
> where the stock price is measured in USD (`Strike`), but the contract pays out in JPY
> (`Collateral`).

Reviewer: ```suggestion
The strike price and collateral often use the same currency. However, this contract
decouples them. You can have, for example, a spread based on APPL stock's price (`Underlying`),
with the stock price in USD (`Strike`), and the contract paying out in JPY
(`Collateral`).
```

> The issuerKeywordRecord specifies the issuers for four keywords: Underlying, Strike, Collateral,
> and Quote. The original deposit and the payout are in Collateral. Strike amounts are used for the
> price oracle's quote as to the value of the Underlying, as well as the strike prices in the
> terms. The priceAuthority's quotes are ERTP assets issued by a QuoteAuthority, which must be
> specified using the keyword Quote.

Reviewer: ```suggestion
The issuerKeywordRecord specifies the issuers for four keywords: 
- Underlying: Strike amounts are used for the price oracle's quote.
- Strike: Strike amounts are used for strike prices in the terms 
- Collateral: Contains the original deposit and the payout.
- Quote: The priceAuthority's quotes are ERTP assets issued by a QuoteAuthority, which must be
  specified using the keyword Quote
```

Reviewer: Honestly, I found the description of the four keywords contents to be very confusing. I'd suggest each one have two distinct parts; what is the value, and where is it coming from/how is it obtained. Definitely want to go with a list, as putting all four and explanations in a single paragraph is harder to follow.

> The terms include { timer, underlyingAmount, expiration, priceAuthority, strikePrice1,
> strikePrice2, settlementAmount }. `timer` must be recognized by `priceAuthority`. `expiration` is a
> time recognized by the `timer`. `underlyingAmount` is passed to `priceAuthority`, so it could be an
> NFT or a fungible amount. `strikePrice2` must be greater than `strikePrice1`. `settlementAmount`
> uses Collateral.

Reviewer: ```suggestion
The terms include:
- `timer`: Must be recognized by `priceAuthority`.
- `underlyingAmount`: Passed to `priceAuthority`, so it could be an NFT or a fungible amount. 
- `expiration`: A time recognized by the `timer`. 
- `priceAuthority`
- `strikePrice1`: Must be less than `strikePrice2`.
- `strikePrice2`: Must be greater than `strikePrice1`.
- `settlementAmount`:  Uses Collateral.
```

Reviewer: Please add something for priceAuthority and expand on how settlementAmount uses Collateral.

<<< @/snippets/zoe/contracts/test-callSpread.js#startInstance

> The terms specify all the details of the options that will be created, but the options are not
> handed out until the creator funds the collateral that will make them valuable.  The
> creatorInvitation has customProperties that include the amounts of the two options as longAmount
> and shortAmount.

Reviewer: ```suggestion
The terms specify all the details of the options that are created. However, the options are not
handed out until the creator funds the collateral that makes them valuable.  The creatorInvitation has 
customProperties including `longAmount` and `shortAmount`,  the amounts of the two options.
```

<<< @/snippets/zoe/contracts/test-callSpread.js#invitationDetails

> The funder can use these option amounts to create an offer that assures them they will get the two
> options in exchange for the funds. The proposal describes the options they want and the collateral
> they give. When the offer is made they get a payout that contains the two option positions, which
> are themselves invitations which can be exercised for free, and provide the option payouts with the
> keyword Collateral.

Reviewer: ```suggestion
The funder can use these option amounts to create an offer assuring they will get the two options in exchange for the funds. The proposal describes the wanted options and the collateral they give. When the offer is made, they get a payout containing the two option positions. The positions are invitations which can be exercised for free, and provide the option payouts with the keyword-specified Collateral.
```

Reviewer: In "When the offer is made, they", "they" is unclear in what it refers to. The use of "a payout" following implies "they" is a singular thing, but since the preceding uses of "they" referred to plurals, you should drop the pronoun and say just who/what is getting the payout (I think it's the funder, since that's the only singular actor given).

Author: I reworded here and changed that to "when the offer is made, a payout is returned".

<<< @/snippets/zoe/contracts/test-callSpread.js#creatorInvitation

> The options that are returned by the contract are Zoe invitations, so their values and terms can be
> verified by asking Zoe for the contract terms.  This makes it possible to sell the options because
> a prospetive purchaser can inspect their value. The prospective purchaser can verify the underlying
> amount, and check to see that the priceAuthority is one they are willing to rely on. They can check
> the expiration to see that it matches their expectations (here it's a small integer suitable for a
> manual timer in a test; in actual use, it would represent block height or wall clock time.) The
> strike prices and settlement amount are likewise visible.

Reviewer: ```suggestion
The options returned by the contract are Zoe invitations, so their values and terms can be
verified by asking Zoe for the contract terms.  This means you can sell the options because
a prospective purchaser can inspect their value by verifying the underlying
amount, and checking that they are willing to rely on this priceAuthority. They can check
that the expiration matches their expectations (here it's a small integer suitable for a
manual timer in a test; in actual use, it represents block height or wall clock time.) The
strike prices and settlement amount are likewise visible.
```

Author: reworded considerably.

<<< @/snippets/zoe/contracts/test-callSpread.js#verifyTerms

The contract doesn't rely on the options being exercised before the specified close. If either

> option is exercised after the closing price is determined, the payouts will still be available. The

Reviewer: ```suggestion
option is exercised after the closing price is determined, the payouts are still available. The
```

Author: This is definitely a statement about the future. Switching to present tense isn't right.

two options make their payment available as soon after exercise as the price is available, and
neither waits for the other to exercise.

<<< @/snippets/zoe/contracts/test-callSpread.js#bobExercise

There is a
[unit test](https://raw.githubusercontent.com/Agoric/agoric-sdk/master/packages/zoe/test/unitTests/contracts/test-callSpread.js)

> (search for "sell options") that demonstrates how these options could be offered using the
> SimpleExchange contract.

Reviewer: ```suggestion
(search for "sell options") showing how you could offer these options using the
SimpleExchange contract.
```

Author: done.

ownership of the underlying asset at the start, and the beneficiaries shouldn't expect to take
delivery at closing.

The Strike and collateral currencies are often the same, however this contract decouples the
currencies. You can have, for example, a spread based on APPL stock (`Underlying`), with the stock
price in USD (`Strike`) and contract paying out in JPY (`Collateral`).

The issuerKeywordRecord specifies the issuers for four keywords: Underlying, Strike, Collateral,
and Quote.

 * The asset whose eventual value determines the payouts uses `Underlying`. This is often a
   fungible currency, but doesn't have to be. It would be perfectly sensible to have a call spread
   contract on the value of a "Superior Magic Sword", as long as there was a price oracle to say
   how its price varies over time.
 * The original deposit and the payout use the `Collateral` issuer.
 * `Strike` amounts are used for the price oracle's quote as to the value of the Underlying, as
   well as the strike prices in the terms.
  * The priceAuthority's quotes are ERTP assets issued by a QuoteAuthority, which must be specified
    using the keyword `Quote`.

> The terms include { timer, underlyingAmount, expiration, priceAuthority, strikePrice1,

Author: I couldn't add one section without labeling what came before and after. I think this is an improvement.
